[Deps] Migrate to Version Catalogs

ParaskP7 opened the pull request:
Project Thread: paaHJt-7cv-p2

This PR migrates all dependencies and plugins in this project to [Version Catalogs](https://docs.gradle.org/current/userguide/platforms.html) and under the newly created [gradle/libs.versions.toml](https://github.com/woocommerce/woocommerce-android/pull/12765/files#diff-697f70cdd88ba88fe77eebda60c7e143f6ad1286bca75017421e93ad84fb87df) file.

-----

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

FYI: I recommend reviewing this PR commit-by-commit just to make sure that every dependency and plugin has been migrated appropriately, just to make sure that there isn't any left-overs that need addressing.

PS: The [gradle/libs.versions.toml](https://github.com/woocommerce/woocommerce-android/pull/12765/files#diff-697f70cdd88ba88fe77eebda60c7e143f6ad1286bca75017421e93ad84fb87df) file, on both, its `versions` and `libraries` sections:
- It is alphabetically ordered.
- It is using the [kebal-case](https://en.wikipedia.org/wiki/Letter_case#Kebab_case) naming convention (or else `dash-case`) .
- All words are lowercased.
- Where it makes sense the `org` name is added in its prefix to group dependencies together and make them easily identifiable.
- Where needed the word `main` is added in its suffix to make sure that a dependency cannot be a parent of another dependency and a dependency itself. This helps in various corner cases.

-----

### Extra Focus Required (⚠️)

Below are some commits you want to put some extra focus on, along with a brief description on each. You can read the full description on the commit itself:

- [libraries]:
    - `google mlkit` -> 9df67a762b1613143c35af6bcf6583189a333cd2
        - `Google MLKit` vs `Google Play Services` dependency naming.
    - `androidx core` -> d967e755d837f837b3a693b4610ded7729f1fb89
        - Upgrade `1.0.0` to `1.0.1` version.
    - `androidx preference` -> 42b9e11bb64b24bf48a1d86380030a7b55877298
        - Upgrade `1.2.0` to `1.2.1` version.
    - `androidx datastore` -> 7f69b0c4c5eb37a934724aa87ae158ab68097b57
        - Upgrade `1.0.0` to  `1.1.0` version.
    - `androidx hilt` -> e810a1f21d4d22566fe3eefdf066fd5d72bec316
        - Downgrade `1.2.0` to `1.1.0` version.
    - `bumptech glide` -> 8893f9ac15c293bfe8206b672af1d23c46ab8359
        - Drop `@aar` from version.
    - `androidx activity` -> 13af208d2e8f0b4597e1bf311a93f542c8e80bd1
        - No associated version for `androidx-activity-compose'`.
    - `androidx compose` -> 9e129c0d92f79c5a0e610d4319d2c3d1a2333366
        - Use `bom` instead of `1.3.2` version.
    - `androidx wear` -> c27691161b777412240b6afb86a614072a472dc1
        - No associated version for `androidx-wear-tiles`.
- [plugins]:
    - `kotlin plugin` -> 73e832f724c4aeb94e7e587c2f0196a7075008ee
        - Drop `org.jetbrains.kotlin.jvm` from plugins.
    - `androidx navigation` -> 5f790e890f9e76ea004a8fe7be305d1b424dc062
        - No associated version for `androidx-navigation-compose`.

On all the above commits I had to make a call, that is, in terms of how to resolve the inconsistency. If any of the changes look suspicious to you, or doesn't seem the right thing to do, please feel free to call me on it and we could revert it (just to keep things as is), or change it to a different resolution (ie. pick a different version).

-----

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

- Smoke test the app and verify everything is working as expected.
- You could also run the `./gradlew buildHealth` task locally and verify that everything is working as expected ([CI](https://buildkite.com/automattic/woocommerce-android/builds/24610)).

-----

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
Deps: Migrate androidx lifecycle to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate androidx test uiautomator to version catalogs on all mods

ParaskP7 committed:
Deps: Migrate junit to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate androidx test ext to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate androidx test main to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate stripe terminal to version catalogs on cardreader

ParaskP7 committed:
Deps: Migrate kotlinx coroutines to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate mockito kotlin to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate mockito inline to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate assertj to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate android billingclient to version catalogs on iap

ParaskP7 committed:
Deps: Migrate androidx appcompat to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate wordpress libaddressinput to version catalogs on wc

ParaskP7 committed:
Deps: Migrate google firebase to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate google play services to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate google mlkit to version catalogs on woocommerce

FYI: Notice that the Japanese, Chinese and Korea 'Text Recognition'
related dependencies are not 'Google MLKit' related but actually
'Google Play Services' related. However, to keep things simple and since
the version is coming from 'Google MLKit' and it main 'Text Recognition'
dependency the overall version catalog is pointing to 'Google MLKit' and
not 'Google Play Services'.

ParaskP7 committed:
Deps: Migrate androidx core to version catalogs on all modules

FYI: Notice that the 'AndroidX Core Splashscreen' version on the
'WooCommerce' module is pointing to '1.0.0', while on the
'WooCommerce-Wear' module it is pointing to '1.0.1'. This change
also updates the 'WooCommerce' version to point to '1.0.1' as well, it
being just a batch update that includes a single bugfix and no more.

Release Notes: https://developer.android.com/jetpack/androidx/releases/
core#core-splashscreen-1.0.1

ParaskP7 committed:
Deps: Migrate androidx constraintlayout to version catalogs on wc

ParaskP7 committed:
Deps: Migrate androidx recyclerview to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate google material to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate androidx transition to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate androidx cardview to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate androidx browser to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate androidx preference to version catalogs on all modules

FYI: Notice that the 'AndroidX Preference KTX' version on the
'WooCommerce' module is pointing to '1.2.0', while on the
'WooCommerce-Wear' module it is pointing to '1.2.1'. This change
also updates the 'WooCommerce' version to point to '1.2.1' as well, it
being just a batch update that includes a few bugfixes and no more.

Release Notes: https://developer.android.com/jetpack/androidx/releases/
preference#1.2.1

ParaskP7 committed:
Deps: Migrate androidx datastore to version catalogs on all modules

FYI: Notice that the 'AndroidX Datastore' versions on the
'WooCommerce' module are pointing to '1.0.0', while on the
'WooCommerce-Wear' module they are pointing to '1.1.0'. This change
also updates the 'WooCommerce' versions to point to '1.1.0' as well, it
having only additional features on top of '1.0.0' but no breaking
changes.

Release Notes: https://developer.android.com/jetpack/androidx/releases/
datastore#1.1.0

ParaskP7 committed:
Deps: Migrate androidx work to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate wordpress utils to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate automattic tracks to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate wordpress fluxc to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate wordpress login to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate wordpress aztec to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate gravatar to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate facebook shimmer to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate photoview to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate automattic about to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate androidx hilt to version catalogs on all modules

FYI: Notice that the 'AndroidX Hilt Compose' version on the
'WooCommerce-Wear' module is pointing to '1.2.0', while on the
'WooCommerce' module it is pointing to '1.1.0', as well as all other
'AndroidX Hilt' dependencies on 'WooCommerce-Wear' too, which too are
pointing to '1.1.0'. This change downgrades the 'WooCommerce-Wear'
version of only 'AndroidX Hilt Compose' to point to '1.1.0' as well,
'1.2.0' has only one important change on top of '1.1.0' but no breaking
changes.

Release Notes: https://developer.android.com/jetpack/androidx/releases/
hilt#hilt_version_12_2

ParaskP7 committed:
Deps: Migrate mpandroidchart to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate bumptech glide to version catalogs on woocommerce

FYI: This change also drops the '@aar' suffix requirement, as the per
the 'Glide v4 Volley' documentation.

PS: The only difference between having the '@aar' suffix and not having
it, is the fact that not having it also bring in transitive
dependencies related to that 'volley-integration' dependency.

ParaskP7 committed:
Deps: Migrate google play to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate google zxing to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate facebook flipper to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate facebook soloader to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate squareup leakcanary to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate androidx arch core to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate androidx test espresso to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate fastlane screengrab to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate wiremock to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate jetty webapp to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate jackson databind to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate json path to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate commons fileupload to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate apache http client android to version catalogs on wc

ParaskP7 committed:
Deps: Migrate zendesk to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate androidx fragment to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate androidx activity to version catalogs on woocommerce

FYI: The 'androidx-activity-compose' dependency has no associated
'androidx-activity' version on it as it seems to be taking this version
from 'androidx-compose-bom' instead.

ParaskP7 committed:
Deps: Migrate cashapp turbine to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate apache commons text to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate commons io to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate tinder statemachine to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate wordpress mediapicker to version catalogs on woocommerce

ParaskP7 committed:
Cleanup: Remove local woocommerce shared path leftover

This is a leftover from this 298efad766dac076a7414fcad34afb978268bf4c
commit that wasn't removed as part of it.

ParaskP7 committed:
Deps: Migrate androidx compose to version catalogs on all modules

FYI: The 'androidx.compose.ui:ui-test-junit4' version is not being
dependant on the 'androidx.compose.bom' version, via this newly added
'androidTestImplementation(platform(libs.androidx.compose.bom))'
configuration, and not hardcoded to the specific '1.3.2' version of this
library.

ParaskP7 committed:
Deps: Migrate coil to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate android desugar to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate androidx camera to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate google guava to version catalogs on woocommerce

ParaskP7 committed:
Deps: Migrate android security lint to version catalogs on all modules

ParaskP7 committed:
Deps: Migrate androidx wear to version catalogs on woocommerce wear

FYI: The 'androidx-wear-tiles' dependencies have no associated
'androidx-wear-tiles' version on them as it seems that they are not to
be taking any version from anywhere.

ParaskP7 committed:
Deps: Migrate google horologist to version catalogs on woocommerce wear

ParaskP7 committed:
Deps: Migrate google gson to version catalogs on woocommerce wear

ParaskP7 committed:
Cleanup: Remove unused dependency versions from root build gradle file

ParaskP7 committed:
Deps: Migrate agp plugin to version catalogs

ParaskP7 committed:
Deps: Migrate google services plugin to version catalogs

ParaskP7 committed:
Deps: Migrate google dagger plugin to version catalogs

ParaskP7 committed:
Deps: Migrate detekt plugin to version catalogs

ParaskP7 committed:
Deps: Migrate kotlin plugin to version catalogs

FYI: The 'org.jetbrains.kotlin.jvm' plugin got completely removed as it
is not being used anywhere.

ParaskP7 committed:
Deps: Migrate ksp plugin to version catalogs

ParaskP7 committed:
Deps: Migrate automattic measure builds plugin to version catalogs

ParaskP7 committed:
Deps: Migrate androidx navigation plugin to version catalogs

FYI: The 'androidx-navigation-compose' dependency on the 'WooCommerce'
module had no associated version on it, and was most probably taking
this version from 'androidx-compose-bom' instead. However, the same
dependency on the 'WooCommerce-Wear' module did have one. This change,
applies the same version on the 'WooCommerce' module to keep things
consistent.

ParaskP7 committed:
Deps: Migrate sentry plugin to version catalogs

ParaskP7 committed:
Deps: Migrate google protobuf plugin to version catalogs

ParaskP7 committed:
Deps: Migrate dependency analysis plugin to version catalogs

ParaskP7 committed:
Deps: Migrate fladle plugin to version catalogs

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/241">20.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on gradle/libs.versions.toml:
```diff
@@ -0,0 +1,249 @@
+androidx-transition = { group = "androidx.transition", name = "transition", version.ref = "androidx-transition" }
+androidx-wear-compose-material = { group = "androidx.wear.compose", name = "compose-material", version.ref = "androidx-wear-compose" }
+androidx-wear-compose-foundation = { group = "androidx.wear.compose", name = "compose-foundation", version.ref = "androidx-wear-compose" }
+androidx-wear-tiles-main = { group = "androidx.wear.tiles", name = "tiles" }
+androidx-wear-tiles-material = { group = "androidx.wear.tiles", name = "tiles-material" }
```
I'd suggest putting an explicit version on this, currently, this dependency is coming as a transitive dependency from the `horologist` dependencies as well, so they dictate the version `1.3.0`, if we remove them, the resolution for the library fails:

<img width="429" alt="Screenshot 2024-10-10 at 12 10 05" src="https://github.com/user-attachments/assets/c53e9d85-9409-4cad-b34b-440c003c59d9">


So I suggest putting the version `1.3.0` to match the existing one, and make it clearer.

ParaskP7 replied:
Great suggestion @hichamboushaba , this is now done: c11e208823a59ab731900c450ab09cc27ccc475a

hichamboushaba commented on build.gradle:
```diff
@@ -3,15 +3,18 @@ import io.gitlab.arturbosch.detekt.Detekt
+    alias(libs.plugins.detekt)
+    alias(libs.plugins.automattic.measure.builds)
+    alias(libs.plugins.dependency.analysis)
+    alias(libs.plugins.android.application).apply(false)
```
very minor np, I think that keeping using ` apply false` instead of `apply(false)` (for example: `alias(libs.plugins.android.application) apply false` feels more familiar to the previous syntax and looks better 😅, I know it's a small thing, so feel free to ignore.

ParaskP7 replied:
Small things matter, this is now done: 970ec5a4257e9dc2e07bdc5bc348f6bef3d54abf

hichamboushaba reviewed (approved):
Excellent work @ParaskP7 👏, this looks really great, and works well (I tested building both the main app and the Wear one).

I left two minor comments, but none is blocking, so I'm pre-approving.

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into deps/migrate-to-version-catalogs

# Conflicts:
#	build.gradle

ParaskP7 committed:
Deps: Update fluxc to 2.99.0

Release Notes: https://github.com/wordpress-mobile/
WordPress-FluxC-Android/releases/tag/2.99.0

The actual update was done as part of
55313270f6245ce94cada95f2f14773fc47ecaa9, but after merging this
branch with 'trunk', to fully resolved this one conflict, the FluxC
update needs to be redone, this time within 'libs.versions.toml'.

ParaskP7 committed:
Deps: Add an explicit version to androidx wear tiles dependencies

As per the c27691161b777412240b6afb86a614072a472dc1 commit, these two
'androidx-wear-tiles' dependencies had no associated version on them.
This could be problematic, non deterministic and could cause trouble
moving forward. As such, and as per the PR comment suggestion below,
this commit adds the most appropriate '1.3.0' version, which is also the
one currently being implicitly set transitively by the 'horologist'
dependencies themselves.

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
12765#discussion_r1795224428

ParaskP7 committed:
Refactor: Revert to previous simpler syntax of alias plugin apply false

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
12765#discussion_r1795251645

ParaskP7 commented:
Thank you so much for the review @hichamboushaba , you rock! 🙇 ❤️ 🚀 

Also, apologies for not coming back to you sooner, being on our meetup last week and all. But, I am back now and ready to pick this up with you again! 💯

ParaskP7 commented:
> Android Lint found more than 20 potential problems in the proposed changes. Check the [Files changed](https://github.com/woocommerce/woocommerce-android/pull/12765/files) tab for more details.

FYI: This is all about the [libs.version.toml](https://github.com/woocommerce/woocommerce-android/pull/12765/files#diff-697f70cdd88ba88fe77eebda60c7e143f6ad1286bca75017421e93ad84fb87df) and the (known) multiple `Obsolete Android Gradle Plugin Version` + `Obsolete Gradle Dependency` warnings on it.

ParaskP7 commented:
👋 @hichamboushaba , based on your comment reactions I think we could now proceed and merge this, right? This is just me making it sure before hitting the button? 😊

hichamboushaba commented:
Yes, you can go ahead @ParaskP7, I didn't merge it myself in case you needed an input from Wojtek as well.

ParaskP7 commented:
Great, right, thanks for the confirmation on that @hichamboushaba ! 🙇 

@wzieba any additional input from you maybe before me merging this? 🙏

wzieba reviewed (approved):
LGTM, thanks for solid work @ParaskP7 !

The dependency diff reports some initially unexpected changes, but after reading the description and commits, things are clear to me 👍 :shipit:

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into deps/migrate-to-version-catalogs

ParaskP7 commented:
Awesome, thanks for looking at this and the 👍 @wzieba ! 🙇  🚀 

FYI: I just merged the latest changes from `trunk` into this PR and upon CI successfully building it I'll merge this. ✅
